### PR TITLE
cpu/stm32: add driver for the LTDC peripheral

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR adds a driver implementation for the LCD-TFT Display Controller (LTDC) available in some STM32 CPUs. A configuration is provided for the STM32F746G-DISCO board and its 4"3inch display.

The driver has basic functionalities and limitations:
- Only one layer is used
- Only 16bit pixel (RGB565) format is supported
- The framebuffer consists in an array in RAM so the driver consumes a lot of RAM on the STM32F746G-DISCO board: 480x272 pixel of 2 bytes each = ~262KB. This limits the driver to single framebuffer rendering. But STM32 has another feature (I'm looking into it) to extend RAM with external SRAM called FMC (Flexible Memory Controller). This can be used to store the framebuffer(s) outside of the internal RAM. I'll open a PR when I'm done with that.

There are a few commits that extend the support for the disp_dev API: the display of the disco boards can be used with existing test applications (with minimal modifications here and there): `tests/pkg_lvgl`, `tests/pkg_qr-code-generator`, `tests/disp_dev`.

Apparently the STM32F429I-DISC1/O boards also have this feature, in follow-up PR I plan to look at that (except if someone else wants to do it) because the rendering is much faster than via the SPI interface of the ILI9341 driver.

### Testing procedure

- Green CI
- `tests/periph_ltdc`, `tests/pkg_lvgl`, `tests/pkg_lvgl_touch`, `tests/pkg_qr-code-generator`, etc are working on STM32F746G-DISCO:

<details>

tests/periph_ltdc

tests/pkg_qr-code-generator

tests/pkg_lvgl_touch

tests/pkg_lvgl

</details>

### Issues/PRs references

None

### CODE REVIEW

File: boards/stm32f746g-disco/board.c [hunk 1]
```diff
@@ -18,9 +18,18 @@
...
 
 void board_init(void)
 {
+    if (IS_USED(MODULE_PERIPH_LTDC)) {
+        gpio_init(LCD_DISP_PIN, GPIO_OUT);
+        gpio_set(LCD_DISP_PIN);
+    }
+    else {
+        gpio_init(BACKLIGHT_PIN, GPIO_OUT);
+        gpio_clear(BACKLIGHT_PIN);
```
Comment (benpicco) [inline boards/stm32f746g-disco/board.c:33]: Is the backlight controlled by  LTDC or why do we only need it if the display is *not* used?

File: boards/stm32f746g-disco/include/periph_conf.h [hunk 2]
```diff
@@ -173,6 +173,57 @@ static const eth_conf_t eth_config = {
...
+        { .pin = GPIO_PIN(PORT_K, 4), .af = GPIO_AF14, },
+        { .pin = GPIO_PIN(PORT_K, 5), .af = GPIO_AF14, },
+        { .pin = GPIO_PIN(PORT_K, 6), .af = GPIO_AF14, },
+    },
+    .hsync      = 41,
+    .vsync      = 10,
+    .hbp        = 13,
+    .hfp        = 32,
+    .vbp        = 2,
+    .vfp        = 2,
```
Comment (benpicco) [inline boards/stm32f746g-disco/include/periph_conf.h:223]: Would be great to have a comment on where those values come from, so people adapting this for their custom board have some pointers.

File: cpu/stm32/include/periph/cpu_ltdc.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+/**
+ * @brief   LTDC GPIO configuration
+ */
+typedef struct {
+    gpio_t pin;                     /**< GPIO pin */
+    gpio_af_t af;                   /**< Alternate function */
+} ltdc_gpio_t;
+
+/**
+ * @brief   Ethernet Peripheral configuration
```
Comment (benpicco) [inline cpu/stm32/include/periph/cpu_ltdc.h:39]:
```suggestion
 * @brief   LTDC Peripheral configuration
```

File: cpu/stm32/periph/ltdc.c [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        gpio_init(ltdc_config.g_pin[pin].pin, GPIO_OUT);
+        gpio_init_af(ltdc_config.g_pin[pin].pin, ltdc_config.g_pin[pin].af);
+        gpio_init(ltdc_config.b_pin[pin].pin, GPIO_OUT);
+        gpio_init_af(ltdc_config.b_pin[pin].pin, ltdc_config.b_pin[pin].af);
+    }
+}
+
+static void _configure_ltdc(void)
+{
+    uint32_t tmp = 0;;
```
Comment (benpicco) [inline cpu/stm32/periph/ltdc.c:89]:
```suggestion
    uint32_t tmp;
```

File: cpu/stm32/periph/ltdc.c [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    /* Set the background color value: black */
+    LTDC->BCCR &= ~(LTDC_BCCR_BCBLUE | LTDC_BCCR_BCGREEN | LTDC_BCCR_BCRED);
+
+    /* Enable LTDC */
+    LTDC->GCR |= LTDC_GCR_LTDCEN;
+}
+
+static void _configure_ltdc_layer(void)
+{
+    uint32_t tmp = 0;;
```
Comment (benpicco) [inline cpu/stm32/periph/ltdc.c:134]:
```suggestion
    uint32_t tmp ;
```

**Review by benpicco — COMMENTED** [review]
Looks good!

File: boards/stm32f746g-disco/board.c (hunk 1 above)
Comment (aabadie) [reply boards/stm32f746g-disco/board.c:33]: That's a small change that indeed turns off the screen if it's unused, saving some power. The backlight is not controller by the LTDC, it's an external pin that is controlled by user code (like here).

File: boards/stm32f746g-disco/include/periph_conf.h (hunk 2 above)
Comment (aabadie) [reply boards/stm32f746g-disco/include/periph_conf.h:223]: I got the value from the Cube code but maybe we can get them from the datasheet of the display module, I'll check

File: boards/stm32f746g-disco/board.c (hunk 1 above)
Comment (aabadie) [reply boards/stm32f746g-disco/board.c:34]: You are right the logic is odd here. Seems like the backlight is enabled by default. `BACKLIGHT_ON/OFF` macros are broken if it's not always initialized. Should be better now.

File: boards/stm32f746g-disco/include/periph_conf.h (hunk 2 above)
Comment (aabadie) [reply boards/stm32f746g-disco/include/periph_conf.h:228]: Seems like it's not that easy, the datasheet gives different values. There was a question about that on ST's forum, I added the link to it.

**Comment by aabadie** [comment]
@benpicco, I think your comments are addressed.

File: boards/stm32f746g-disco/include/periph_conf.h (hunk 2 above)
Comment (benpicco) [reply boards/stm32f746g-disco/include/periph_conf.h:228]: Then just add a comment saying they are generated by CubeMX, so someone who wants to implement this for their board knows where to start looking

**Comment by benpicco** [comment]
Please squash!

**Review by benpicco — APPROVED** [review]

**Pull request merged**